Use abort instead of exit(2) in caml_fatal_error.

jhjourdan opened the pull request:
This makes it possible to get a core dump in case of a fatal error, which makes debugging easier.

This has been suggested in #8628.

jhjourdan committed:
Use abort instead of exit(2) in caml_fatal_error.

This makes it possible to get a core dump in case of a fatal error,
which makes debugging easier.

This also
- turns a use of exit(2) to call to caml_fata_error
- replaces some uses of caml_fatal_error to a call to exit(127), when
  the error results in misuse of the runtime
- removes some useless calls to exit(-1)

dbuenzli commented:
A `git grep 'exit('` in `runtime` shows there are a few other places in the runtime system where it `exit(2)`. I think at least one of these can safely be turned into `caml_fatal_error`: it's the first hit in `win32.c`. I'm unclear about those in `fail_{byt,nat}.c`.

The other ones and those calling `caml_sys_exit` are more tricky since they also involve `Stdlib.at_exit` handlers which you might not expect to run before an `abort()`. In particular changing the default behaviour in case of uncaught exception in `printexc.c` might not be a great idea and this one the user program can at least prevent it from happening and messing with its own `exit` usage.

jhjourdan commented:
Well spotted! I turned calls to exit(2) in `fail_xxx.c` and `win32.c` into fatal errors. I also fixed one typo and removed spurious calls to `exit(-1)` in the debugger.

stedolan commented:
I think this is a good idea: when the OCaml runtime fails, it is better to `abort()` than to exit with an error code.

However, `caml_fatal_error` is also used in the bytecode runtime for several more boring errors (e.g. if you run `ocamlrun file_that_does_not_exist`, you get a fatal error). I don't think these situations should trigger a SIGABRT.

I think the right thing to do here is to have `caml_fatal_error` call abort, but call `exit(N)` instead in these situations in `runtime/startup_byt.c`. I'm not sure what the exit code should be: ocaml currently exits with 2, but I think 127 is a more normal code for when an executable file cannot be found.

jhjourdan commented:
> I think this is a good idea: when the OCaml runtime fails, it is better to `abort()` than to exit with an error code.
> 
> However, `caml_fatal_error` is also used in the bytecode runtime for several more boring errors (e.g. if you run `ocamlrun file_that_does_not_exist`, you get a fatal error). I don't think these situations should trigger a SIGABRT.
> 
> I think the right thing to do here is to have `caml_fatal_error` call abort, but call `exit(N)` instead in these situations in `runtime/startup_byt.c`. I'm not sure what the exit code should be: ocaml currently exits with 2, but I think 127 is a more normal code for when an executable file cannot be found.

Done.

xavierleroy commented on runtime/fail_byt.c:
```diff
@@ -93,18 +93,14 @@ CAMLexport void caml_raise_with_string(value tag, char const *msg)
-    fprintf(stderr, "Fatal error: exception %s\n", exception_name);
-    exit(2);
-  }
+  if (caml_global_data == 0)
```
This code is entered when a C function raises an exception and the bytecode runtime system is not initialized yet, i.e. the table of globals is not available.  This can happen in mixed C/OCaml programs that start in C then call into OCaml.  

The previous behavior was to treat this situation like an uncaught exception --- which it is, in a way: an exception is raised (from C) and no handler (in OCaml) is there to catch it.  Hence the error message and exit code were the same than those for an uncaught exception.  

Now you're turning this into an abort() with core dump.  This is not the same behavior.

jhjourdan replied:
What you are saying somehow contradicts with the comment above this part of code: I thought the `caml_raise_xxx` functions were not meant to be used when the runtime is not initialized (like anything from the OCaml runtime, BTW). Rather, as the comment above suggests, I thought this check is needed if the `intern` function itself (which is responsible for decoding global data) throws an exception.

If it is only about the exceptions raised by `intern`, then I would argue that `abort` is the right thing to do here. Indeed, the only case I can see that would explain this happening in practice is a malformed bytecode executable (or under extreme memory constrains), which, in a way, is an internal error of the OCaml system (i.e., `ocamlc` produced a wrong executable).

xavierleroy commented on runtime/fail_byt.c:
```diff
@@ -93,18 +93,14 @@ CAMLexport void caml_raise_with_string(value tag, char const *msg)
-    fprintf(stderr, "Fatal error: exception %s(\"%s\")\n", exception_name, msg);
-    exit(2);
-  }
+  if (caml_global_data == 0)
```
Same comment as above.

xavierleroy commented on runtime/fail_nat.c:
```diff
@@ -176,11 +177,8 @@ void caml_array_bound_error(void)
-                      "Invalid_argument(\"index out of bounds\")\n");
-      exit(2);
-    }
+    if (caml_array_bound_error_exn == NULL)
```
Same thing here: `caml_array_bound_error_exn` can be NULL if the OCaml code hasn't been initialized  yet, and the C code raises the out-of-bounds exception.  

The previous code was careful to report this situation like an uncaught exception `Invalid_argument("index out of bounds")`, and the new code just dumps core.

jhjourdan replied:
Same argument as above: if we assume that the C code is not calling OCaml runtime functions before initializing that runtime, then this can only happen if the standard library failed to initialize this named value. I consider this as an internal OCaml error.

The assumption that the client code should not call runtime function before initialization is reasonable: you don't want to check in e.g., `caml_alloc` that the runtime is initialized, do you?

xavierleroy commented on runtime/startup_byt.c:
```diff
@@ -375,27 +376,32 @@ CAMLexport void caml_main(char_os **argv)
-      caml_fatal_error("no bytecode file specified");
+    if (argv[pos] == 0) {
+      fprintf(stderr, "no bytecode file specified");
+      exit(127);
```
This exit code 127 is a good idea.

xavierleroy reviewed (commented):
I'm OK with the general idea: an internal error in the runtime system should abort() rather than exit(2).

However,  not all calls to `caml_fatal_error` represent internal errors in the runtime system.  Some are errors in the OCaml program being executed, and should not abort().  

One instance is when the bytecode executable cannot be found, and you already took care of that.  

Another instance, detailed below, is when the C code tries to raise an OCaml exception and the OCaml code isn't initialized yet.  This case was carefully reported like an uncaught exception, and you turned it into an abort().  This is debatable.

jhjourdan commented:
> Another instance, detailed below, is when the C code tries to raise an OCaml exception and the OCaml code isn't initialized yet. This case was carefully reported like an uncaught exception, and you turned it into an abort(). This is debatable.

As I explained, I think it should be considered UB to call OCaml runtime functions if said runtime is not initialized. Most of the external functions of the OCaml runtime requires it to be initialized anyway (allocation...). I think the test in `check_global_data` is only there because `intern` can raise an exception while decoding the bytecode executable.

If you don't agree, then I'll be happy to go back to the previous version (which is just a little less factorized). I just need to know what you think to move forward :)

ygrek commented:
> I'm OK with the general idea: an internal error in the runtime system should abort() rather than exit(2).

ftr now I strongly support @dbuenzli idea of overridable C handler for this kind of unexpected situations. E.g. in our usecase we want to output log in some specific format so that our monitoring tools can track this situation same way as all other logs. Also from practice - we may not always want to do a coredump on oom (this contradicts what I said before, but that's what practice taught me - consuming all IO of a rood disk to get a 200G coredump is not always desirable).

xavierleroy commented:
The size of core dumps is not a factor here: just use `ulimit -c N` to limit the max size of core dumps, or `ulimit -c 0` to make sure that core dumps are never generated.

This said, I am sympathetic to the idea that users could provide their own handler function instead of the default `abort()`.

jhjourdan commented:
> ftr now I strongly support @dbuenzli idea of overridable C handler for this kind of unexpected situations. 

Isn't it possible to install a signal handler for SIGABRT?

xavierleroy commented:
> As I explained, I think it should be considered UB to call OCaml runtime functions if said runtime is not initialized.

This is a reasonable claim that is open for discussion.  I still think you need to know why the code you're modifying is written the way it is written, and be aware of what you are changing, rather than changing it without understanding it.

ygrek commented:
> The size of core dumps is not a factor here: just use ulimit -c N to limit the max size of core dumps, or ulimit -c 0 to make sure that core dumps are never generated.

We still want to get (full) coredumps for real segfaults..

> Isn't it possible to install a signal handler for SIGABRT?

maybe, but I think explicit handler is better mechanism

dbuenzli commented:
> Isn't it possible to install a signal handler for SIGABRT?

In the handler the context is lost, is the abort coming from OCaml or from C ? If my main is in C I may simply choose to shutdown the OCaml system and restart it.

I think it's a good idea if the OCaml runtime strives to be treated as C library. In other words that it tries as much as possible not to fiddle with the non compositional aspects of the C runtime/standard library.

jhjourdan committed:
Add [caml_fatal_error_hook].

When not NULL, this hook is called when a fatal error is encountered
instead of printing an error message on stderr.

jhjourdan commented:
@xavierleroy : alright, I restored the initial code for the three changes you mentioned. 

@dbuenzli , @ygrek : your wish is granted. `caml_fatal_error` now calls `caml_fatal_error_hook` if not `NULL`. Note, however, that this hook is not called in case of unhandled exception or failed startup.

xavierleroy reviewed (approved):
The diff is getting shorter, I like it :-)    More seriously: this looks very good to me.

xavierleroy commented:
Do I understand correctly that the hook mechanism answers @ygrek's concerns about core dumps?  Or is there still a problem with standalone applications?

ygrek commented on Changes:
```diff
@@ -29,6 +29,9 @@ Working version
    review by Xavier Leroy and Gabriel Scherer)
 
+- #8630: Use abort() instead of exit(2) in caml_fatal_error.
+  (Jacques-Henri Jourdan, review by Xavier Leroy)
```
please mention new hook in changes

ygrek reviewed (approved):
I believe this covers all my concerns, yes

jhjourdan commented:
I rebased, squashed the commits, and mentionned the new hook in Chances. @xavierleroy : if this looks good to you , then this should be ready to merge.

jhjourdan committed:
Merge branch 'trunk' into fatal_error_abort

jhjourdan commented:
@xavierleroy: you approved these changes one 1.5 months ago. Is there anything blocking the merge?

xavierleroy committed:
Merge branch 'trunk' into fatal_error_abort

xavierleroy commented:
I guess I was hoping someone else would bother to merge :-)  

Let's merge, then.
